.tools/validate.go

```diff
@@ -0,0 +1,269 @@
+	Body                 string
+	CommitNotes          string `json:"commit_notes"`
+	VerificationFlag     string `json:"verification_flag"`
+	ShortMsg             string
+	Signer               string
+	SignerKey            string       `json:"signer_key"`
+	Author               PersonAction `json:"author,omitempty"`
+	Commiter             PersonAction `json:"commiter,omitempty"`
+}
+
+// PersonAction is a time and identity of an action on a git commit
+type PersonAction struct {
+	Name  string `json:"name"`
+	Email string `json:"email"`
+	Date  string `json:"date"` // this could maybe be an actual time.Time
+}
+
+var (
+	prettyLogSubject       = `--pretty=format:%s`
+	prettyLogBody          = `--pretty=format:%b`
+	prettyLogCommit        = `--pretty=format:%H`
+	prettyLogAuthorName    = `--pretty=format:%aN`
+	prettyLogAuthorEmail   = `--pretty=format:%aE`
+	prettyLogCommiterName  = `--pretty=format:%cN`
+	prettyLogCommiterEmail = `--pretty=format:%cE`
+	prettyLogFormat        = `--pretty=format:{"commit": "%H", "abbreviated_commit": "%h", "tree": "%T", "abbreviated_tree": "%t", "parent": "%P", "abbreviated_parent": "%p", "refs": "%D", "encoding": "%e", "sanitized_subject_line": "%f", "commit_notes": "%N", "verification_flag": "%G?", "signer": "%GS", "signer_key": "%GK", "author": { "date": "%aD" }, "commiter": { "date": "%cD" }}`
+)
+
+// GitLogCommit assembles the full information on a commit from its commit hash
+func GitLogCommit(commit string) (*CommitEntry, error) {
+	buf := bytes.NewBuffer([]byte{})
+	cmd := exec.Command("git", "log", "-1", prettyLogFormat, commit)
+	cmd.Stdout = buf
+	cmd.Stderr = os.Stderr
+
+	if err := cmd.Run(); err != nil {
+		log.Println(strings.Join(cmd.Args, " "))
+		return nil, err
+	}
+	c := CommitEntry{}
+	output := buf.Bytes()
+	if err := json.Unmarshal(output, &c); err != nil {
+		fmt.Println(string(output))
+		return nil, err
+	}
+
+	output, err := exec.Command("git", "log", "-1", prettyLogSubject, commit).Output()
+	if err != nil {
+		return nil, err
+	}
+	c.Subject = strings.TrimSpace(string(output))
+
+	output, err = exec.Command("git", "log", "-1", prettyLogBody, commit).Output()
+	if err != nil {
+		return nil, err
+	}
+	c.Body = strings.TrimSpace(string(output))
+
+	output, err = exec.Command("git", "log", "-1", prettyLogAuthorName, commit).Output()
+	if err != nil {
+		return nil, err
+	}
+	c.Author.Name = strings.TrimSpace(string(output))
+
+	output, err = exec.Command("git", "log", "-1", prettyLogAuthorEmail, commit).Output()
+	if err != nil {
+		return nil, err
+	}
+	c.Author.Email = strings.TrimSpace(string(output))
+
+	output, err = exec.Command("git", "log", "-1", prettyLogCommiterName, commit).Output()
+	if err != nil {
+		return nil, err
+	}
+	c.Commiter.Name = strings.TrimSpace(string(output))
+
+	output, err = exec.Command("git", "log", "-1", prettyLogCommiterEmail, commit).Output()
+	if err != nil {
+		return nil, err
+	}
+	c.Commiter.Email = strings.TrimSpace(string(output))
```

Comment:
Not without reflect, which is overkill for doing a simple loop.
On Sep 9, 2015 6:02 PM, "W. Trevor King" <notifications@github.com> wrote:

> In .tools/validate.go
> <https://github.com/opencontainers/specs/pull/167#discussion_r39102627>:
>
> > +	if err != nil {
> > +		return nil, err
> > +	}
> > +	c.Author.Email = strings.TrimSpace(string(output))
> > +
> > +	output, err = exec.Command("git", "log", "-1", prettyLogCommiterName, commit).Output()
> > +	if err != nil {
> > +		return nil, err
> > +	}
> > +	c.Commiter.Name = strings.TrimSpace(string(output))
> > +
> > +	output, err = exec.Command("git", "log", "-1", prettyLogCommiterEmail, commit).Output()
> > +	if err != nil {
> > +		return nil, err
> > +	}
> > +	c.Commiter.Email = strings.TrimSpace(string(output))
>
> Does Go not have an equivalent of Python's setattr? In Python, I'd make
> this:
>
> for format, keys in [
>         ('%s', ['subject']),
>         …
>         ('%cE', ['committer', 'email']),
>         ]:
>     value = …
>     obj = commit
>     for key in keys[:-1]:
>         obj = getattr(obj, key)
>     setattr(obj, keys[-1], value)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/opencontainers/specs/pull/167/files#r39102627>.
>